arwen/contexts/async_local.go

```diff
@@ -0,0 +1,454 @@
+package contexts
+
+import (
+	"math/big"
+
+	"github.com/ElrondNetwork/arwen-wasm-vm/v1_4/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/v1_4/math"
+	"github.com/ElrondNetwork/elrond-go-core/data/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func (context *asyncContext) executeAsyncLocalCalls() error {
+	for {
+		call := context.getNextLocalAsyncCall()
+		if call == nil {
+			break
+		}
+
+		err := context.executeAsyncLocalCall(call)
+		if err != nil {
+			return err
+		}
+
+		context.closeCompletedAsyncCalls()
+	}
+
+	return nil
+}
+
+func (context *asyncContext) executeCompletedGroupCallbacks() {
+	for _, group := range context.asyncCallGroups {
+		if group.IsComplete() {
+			context.executeCallGroupCallback(group)
+		}
+	}
+
+}
+
+func (context *asyncContext) getNextLocalAsyncCall() *arwen.AsyncCall {
+	for _, group := range context.asyncCallGroups {
+		for _, call := range group.AsyncCalls {
+			if call.IsLocal() {
+				return call
+			}
+		}
+	}
+
+	return nil
+}
+
+func (context *asyncContext) executeAsyncLocalCall(asyncCall *arwen.AsyncCall) error {
+	if asyncCall.ExecutionMode == arwen.ESDTTransferOnCallBack {
+		context.executeESDTTransferOnCallback(asyncCall)
+		return nil
+	}
+
+	destinationCallInput, err := context.createContractCallInput(asyncCall)
+	if err != nil {
+		return err
+	}
+
+	// Briefly restore the AsyncCall GasLimit, after it was consumed in its
+	// entirety by addAsyncCall(); this is required, because ExecuteOnDestContext()
+	// must also consume the GasLimit in its entirety, before starting execution,
+	// but will restore any GasRemaining to the current instance.
+	metering := context.host.Metering()
+	metering.RestoreGas(asyncCall.GetGasLimit())
+
+	vmOutput, err := context.host.ExecuteOnDestContext(destinationCallInput)
+
+	// The vmOutput instance returned by host.ExecuteOnDestContext() is never nil,
+	// by design. Using it without checking for err is safe here.
+	asyncCall.UpdateStatus(vmOutput.ReturnCode)
+
+	if asyncCall.HasCallback() {
+		callbackVMOutput, callbackErr := context.executeSyncCallback(asyncCall, vmOutput, err)
+		context.finishAsyncLocalExecution(callbackVMOutput, callbackErr)
+		context.gasAccumulated = 0
+		if callbackVMOutput != nil {
+			context.accumulateGas(callbackVMOutput.GasRemaining)
+		}
+	} else {
+		context.gasAccumulated = 0
+		if vmOutput != nil {
+			context.accumulateGas(vmOutput.GasRemaining)
+		}
+	}
+
+	return nil
+}
+
+func (context *asyncContext) executeESDTTransferOnCallback(asyncCall *arwen.AsyncCall) {
+	context.host.Output().PrependFinish(asyncCall.Data)
+
+	// TODO discuss: the contract has already paid the gas for GasLimit and
```

Comment:
yes. the gas for the transfer was already paid for.

Reply:
Deleted the `TODO`, left the comment for future information.